# program solving and programming


### Day 7
### Date-10th may-19



# problem1:
### problem statement
Procedure to count the number of digits in a given number


 ### constraints
 given number will be in range[1,10000000000]
 

### test cases
* CountDigits(123456)->6
* CountDigits (0)->1



In [3]:
def countdigits(n):
    count=0
    while(n>0):
        count=count+1
        n=n//10
        
    return(count)
        
print(countdigits(123456))
print(countdigits(0))

6
0


# problem2:
### problem statement
Design a procedure to calculate the squareroot of a number "without using the math function sqrt


 ### constraints
 given number will be in range[1,10000000000]
 

### test cases
* Squareroot(25)->5
* Squareroot(36)->6
* Squareroot(100)->6


In [4]:
def squareroot(n):
    for i in range(1,n**2):
        s=pow(n,0.5)
    return s
print(squareroot(25))
print(squareroot(36))
print(squareroot(100))

5.0
6.0
10.0


# problem3:
### problem statement
Design a procedure to sort a given array in ascending order


 ### constraints
 given number will be in range[1,10000000000]
 

### test cases
* sort([3, 2, 1, 4, 4, 5, 6])->[1, 2, 3, 4, 4, 5, 6]
* sort([1, 2, 3])->[1, 2, 3]

In [6]:
def sorting(n):
    n.sort()
    return n
print(sorting([3,2,1,4,4,5,6]))
print(sorting([1,2,3]))

[1, 2, 3, 4, 4, 5, 6]
[1, 2, 3]


# problem 4:
### problem statement
Procedure to check if a given number is Prime and returns a Boolean value.


 ### constraints
 given number will be in range[1,10000000000,boolean]
 

### test cases
* isprime(7)->true
* isprime(9)->false

In [47]:
def isprime(n):
    count=0
    for i in range(2,n):
        if(n%i==0):
            count +=1
    if (count==0):
        return True
    else:
        return False
print(isprime(7))
print(isprime(9))

True
False


# problem5:
### problem statement
Design a flowchart to evaluate the polynomial
f(x)= x^3 + 2 X^2 + 3 X - 10 for x in the range [1, 1000]


 ### constraints
 given number will be in range[1,10000000000]
 

### test cases

In [22]:
poly=int(input("enter a number:"))
print(poly**3+2*poly**2+3*poly-10)

enter a number:720
374286950


# problem 6:
### problem statement
power of anumber using recursion


 ### constraints
 given number will be in range[1,10000000000]
 

### test cases
* recursivepower(2,3)->8
* recursivepower(2,10)->1024

In [7]:
def recursivepower(n,r):
    if r==1:
        return n
    else:
        return recursivepower(n,r-1) * n
print(recursivepower(2,3)) 
print(recursivepower(2,10))

8
1024


# problem 7:
### problem statement
factorial of a number using recursion


 ### constraints
 given number will be in range[1,10000000000]
 

### test cases
* factorial(5)->120
* factorial(6)->720
* factroial(7)->5040

In [11]:
def factorial(n):
    if n==1:
        return n
    else:
        return factorial(n-1)*n
print(factorial(5))
print(factorial(6))
print(factorial(7))

120
720
5040


# problem 8:
### problem statement
gcd using recursion and iterative


 ### constraints
 given number will be in range[1,10000000000]
 

### test cases
gcdrecursive(12,39)->3

In [15]:
#recursive gcd
#gcd of two numbers remains the same when 
#smaller parameter is reduced to the remainder
#and larger parameter is replaced by smaller number
#gcd(a,b)gcd(b%a,a)
def gcdrecursive(a,b):
    if b%a==0:
        return a
    else:
        return gcdrecursive(b%a,a)
gcdrecursive(12,39)    

3

In [20]:
#iterative gcd
def gcditerative(a,b):
    r=b%a
    if (r==0):
        return a
    while r!=0:
        r=b%a
        b=a
        a=r
    return b
gcditerative(6,12)

6

# problem 9:
### problem statement
tower of hanoi using recursion


 ### constraints
 given number will be in range[1,10000000000]
 

### test cases


### explanation
n disks->move n-1 from source to temporary

         move nth disk from source to destination
         
         move n-1 disk fromtemporary to destination
3 disks->move 2nd disk from source to temporary

         move 3rd disk from source to destination
         
         move 2nd from temporary to destination

In [23]:
def towersofhanoi(n,source,temp,dest):
    if n==1:
        print('move ',n,'disk from ',source,'to',dest)
        return
    else:
        towersofhanoi(n-1,source,dest,temp)
        print('move ',n,'disk from ',source,'to',dest)
        towersofhanoi(n-1,temp,source,dest)
        return
towersofhanoi(3,'a','b','c')

move  1 disk from  a to c
move  2 disk from  a to b
move  1 disk from  c to b
move  3 disk from  a to c
move  1 disk from  b to a
move  2 disk from  b to c
move  1 disk from  a to c


# problem 10:
### problem statement
genarating the nth fibonacci number using recursion


 ### constraints
 given number will be in range[1,10000000000]
 

### test cases
* fibonacci(2)->1
* fibonacci(7)->13
* fibonacci(0)->0


In [49]:
def fibonacci(n):
    if n==0:
        return 0

    elif n==1:
        return 1
    elif n==2:
        return 1
    else:
        return (fibonacci(n-1)+fibonacci(n-2))
    
print(fibonacci(2))
print(fibonacci(7))
print(fibonacci(0))
    

1
13
0


# problem 11:
### problem statement
Define a function to convert a decimal number to the corresponding binary number

 ### constraints
 given number will be in range[1,10000000000]
 

### test cases
* 
* 


In [55]:
def decimaltobinary(n):
    if(n>1):
        decimaltobinary(n//2)
    print(n%2,end='') 
    
decimaltobinary(15)


1111